File: 2020/7.py
bags = {}
with open("7.txt") as file:
    for line in file:
        info = line.strip().split(" bags contain ")
        contents = info[1].split(", ")
        if not info[0] in bags:
            bags[info[0]] = contents

for k, v in bags.items():
    if v[0] == "no other bags":
        bags[k] = 0
    else:
        d = {}
        for bag in v:
            params = bag.split(" ")
            d[params[1] + " " + params[2]] = int(params[0])
        bags[k] = d

goldenBags = []
for k, v in bags.items():
    if v != 0 and ("shiny gold" in v):
        goldenBags.append(k)

while True:
    check = len(goldenBags)
    for k, v in bags.items():
        if v != 0:
            for goldenBag in goldenBags:
                if goldenBag in v and not k in goldenBags:
                    goldenBags.append(k)
    if len(goldenBags) == check: 
        break
print(len(goldenBags))

bagsInGold = {}
for k, v in bags.items():
    if k == "shiny gold":
        for bag in v:
            if not bag in bagsInGold:
                bagsInGold[bag] = 0

while True:
    check = len(bagsInGold)
    for k, v in bags.items():
        if k in bagsInGold and v != 0:
            for bag in v:
                if not bag in bagsInGold:
                    bagsInGold[bag] = 0
        elif v == 0:
            if not k in bagsInGold:
                bagsInGold[k] = 0
    if check == len(bagsInGold):
        break

for bag, amount in bags["shiny gold"].items():
    if bag in bagsInGold:
        bagsInGold[bag] = bagsInGold[bag] + amount

for bag, total in bagsInGold.items():
    if bags[bag] == 0:
        continue
    else:    
        for bag, amount in bags[bag].items():
            bagsInGold[bag] = bagsInGold[bag] + total*(amount)

total = 0
for _, value in bagsInGold.items():
    total += value
print(total)

File: 2020/Lua/7.lua
local bags = {}
for line in io.lines("7.txt") do
  local color, contents = line:match("(.+) bags contain (.+)%.")
  if contents ~= "no other bags" then
    local contains = {}
    for c in contents:gmatch("%d+ .- bags?,?") do
      local cont, col = c:match("(%d+) (.-) bags?,?")
      contains[col] = cont
    end
    bags[color] = contains
  end
end

local function getLen(tbl)
  local count = 0
  for _ in pairs(tbl) do count = count + 1 end return count end

local golden = {}
while true do
  local check = getLen(golden)
  for bag, conts in pairs(bags) do
    if conts["shiny gold"] then
      golden[bag] = true
    end
    for k in pairs(conts) do
      if golden[k] then
        golden[bag] = true
      end
    end
  end
  if getLen(golden) == check then print(check) break end
end

local count = 0

local function getBags(inp)
  local contents = {}
  for bag, n in pairs(inp) do
    if bags[bag] then
      for cont, amount in pairs(bags[bag]) do
        if contents[cont] then contents[cont] = contents[cont] + n * amount else contents[cont] = n * amount end
        count = count + n * amount
      end
    end
  end
  return contents
end

local contents = {["shiny gold"] = 1}
while true do
  local check = getLen(contents)
  if check == 0 then print(count) break
else
  contents = getBags(contents)
  end
end
